File: ExelBidDemoObjC/ExelBidDemoObjC/Mediation/MediatedBannerExampleViewController.m
#import "MediatedBannerExampleViewController.h"
#import "DesignSystem.h"
#import "StatusBadge.h"
#import "AdUnitIds.h"
@import ExelBidSDK;

/// Mediated 320x50 banner. ObjC demo omits the waterfall log card —
/// `EBWaterfallEvent` is a Swift enum and isn't bridged to ObjC.
@interface MediatedBannerExampleViewController ()
@property (nonatomic, strong) EBStatusBadge *badge;
@property (nonatomic, strong) EBInfoRow *winnerRow;
@property (nonatomic, strong) EBLogView *logView;
@property (nonatomic, strong) UIView *bannerHost;
@property (nonatomic, strong, nullable) EBMediatedBannerAd *bannerAd;
@end

@implementation MediatedBannerExampleViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.systemGroupedBackgroundColor;
    self.title = @"Mediated Banner";

    UIScrollView *scroll = [[UIScrollView alloc] init];
    scroll.translatesAutoresizingMaskIntoConstraints = NO;
    scroll.alwaysBounceVertical = YES;
    [self.view addSubview:scroll];

    UIStackView *stack = [[UIStackView alloc] init];
    stack.axis = UILayoutConstraintAxisVertical;
    stack.spacing = EBSpacingL;
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    [scroll addSubview:stack];

    EBCardView *intro = [[EBCardView alloc] init];
    [intro addArranged:[EBSectionLabel make:@"Mediated Banner"]];
    [intro addArranged:[EBTypography body:@"EBMediatedBannerAd — 서버가 정한 순서대로 어댑터를 순차 시도하고 첫 성공 네트워크의 광고를 노출합니다. autoRefresh는 지원되지 않습니다."]];
    [stack addArrangedSubview:intro];

    EBCardView *status = [[EBCardView alloc] init];
    [status addArranged:[EBSectionLabel make:@"Status"]];
    self.badge = [[EBStatusBadge alloc] initWithFrame:CGRectZero];
    [status addArranged:self.badge];
    self.winnerRow = [[EBInfoRow alloc] initWithKey:@"Winning network" value:@"-"];
    [status addArranged:self.winnerRow];
    [stack addArrangedSubview:status];

    EBCardView *actions = [[EBCardView alloc] init];
    [actions addArranged:[EBSectionLabel make:@"Action"]];
    [actions addArranged:[EBPrimaryButton filled:@"Load mediated banner" target:self action:@selector(handleLoad)]];
    [stack addArrangedSubview:actions];

    EBCardView *hostCard = [[EBCardView alloc] init];
    [hostCard addArranged:[EBSectionLabel make:@"Creative"]];
    self.bannerHost = [[UIView alloc] init];
    self.bannerHost.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [self.bannerHost.heightAnchor constraintEqualToConstant:50]
    ]];
    [hostCard addArranged:self.bannerHost];
    [stack addArrangedSubview:hostCard];

    EBCardView *logCard = [[EBCardView alloc] init];
    [logCard addArranged:[EBSectionLabel make:@"Lifecycle log"]];
    self.logView = [[EBLogView alloc] initWithFrame:CGRectZero];
    [logCard addArranged:self.logView];
    [stack addArrangedSubview:logCard];

    [NSLayoutConstraint activateConstraints:@[
        [scroll.topAnchor      constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [scroll.bottomAnchor   constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
        [scroll.leadingAnchor  constraintEqualToAnchor:self.view.leadingAnchor],
        [scroll.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [stack.topAnchor      constraintEqualToAnchor:scroll.contentLayoutGuide.topAnchor constant:EBSpacingL],
        [stack.bottomAnchor   constraintEqualToAnchor:scroll.contentLayoutGuide.bottomAnchor constant:-EBSpacingXXL],
        [stack.leadingAnchor  constraintEqualToAnchor:scroll.frameLayoutGuide.leadingAnchor constant:EBSpacingL],
        [stack.trailingAnchor constraintEqualToAnchor:scroll.frameLayoutGuide.trailingAnchor constant:-EBSpacingL]
    ]];

    [self.badge setState:EBStatusBadgeStateIdle text:@"Tap to load"];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.bannerAd stop];
}

- (void)handleLoad {
    EBMediatedBannerAd *ad = self.bannerAd ?: [self makeBannerAd];
    [ad stop];
    [self.winnerRow setValue:@"-"];
    [self setState:EBStatusBadgeStateLoading text:@"loading…"];
    [ad load];
}

- (EBMediatedBannerAd *)makeBannerAd {
    EBMediatedBannerAd *banner = [[EBMediatedBannerAd alloc] initWithAdUnitId:AdUnitIds.banner
                                                                        size:CGSizeMake(320, 50)];
    banner.perNetworkTimeout = 5.0;
    __weak typeof(self) weakSelf = self;
    banner.rootViewControllerProvider = ^UIViewController * _Nullable { return weakSelf; };
    banner.translatesAutoresizingMaskIntoConstraints = NO;
    [self.bannerHost addSubview:banner];
    [NSLayoutConstraint activateConstraints:@[
        [banner.centerXAnchor constraintEqualToAnchor:self.bannerHost.centerXAnchor],
        [banner.centerYAnchor constraintEqualToAnchor:self.bannerHost.centerYAnchor],
        [banner.widthAnchor   constraintEqualToConstant:320],
        [banner.heightAnchor  constraintEqualToConstant:50]
    ]];

    __weak typeof(banner) weakBanner = banner;
    banner.onLoad = ^{
        [weakSelf.winnerRow setValue:weakBanner.winningNetwork ?: @"-"];
        [weakSelf setState:EBStatusBadgeStateReady text:@"loaded"];
    };
    banner.onFailureBlock = ^(NSError *error) {
        [weakSelf setState:EBStatusBadgeStateFailed
                      text:[NSString stringWithFormat:@"failed: %@", error.localizedDescription]];
    };
    banner.onClick        = ^{ [weakSelf setState:EBStatusBadgeStateClicked text:@"clicked"]; };
    banner.onLeaveApp     = ^{ [weakSelf setState:EBStatusBadgeStateClicked text:@"left app"]; };
    banner.onClickFinish  = ^{ [weakSelf setState:EBStatusBadgeStateReady   text:@"click finished"]; };
    self.bannerAd = banner;
    return banner;
}

- (void)setState:(EBStatusBadgeState)state text:(NSString *)text {
    [self.badge setState:state text:text];
    [self.logView append:text];
}

@end

File: ExelBidDemo/ExelBidDemo/Mediation/MediatedBannerExampleViewController.swift
import UIKit
import ExelBidSDK

/// Mediated 320×50 banner. Mirrors `BannerExampleViewController`, swapping
/// `EBBannerAd` for `EBMediatedBannerAd` and adding a dedicated waterfall
/// log card so per-network attempts (try → won/lost → noFill) are visible.
final class MediatedBannerExampleViewController: UIViewController {

    private let badge = StatusBadge()
    private let winnerRow = InfoRow(key: "Winning network", value: "-")
    private let logView = LogView()
    private let waterfallView = LogView()
    private let options = TestOptionsPanel()
    private let bannerHost = UIView()

    private var bannerAd: EBMediatedBannerAd?

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .systemGroupedBackground
        title = "Mediated Banner"

        let scroll = UIScrollView()
        scroll.translatesAutoresizingMaskIntoConstraints = false
        scroll.alwaysBounceVertical = true
        view.addSubview(scroll)

        let stack = UIStackView()
        stack.axis = .vertical
        stack.spacing = Spacing.l
        stack.translatesAutoresizingMaskIntoConstraints = false
        scroll.addSubview(stack)

        let intro = CardView()
        intro.addArranged(SectionLabel.make("Mediated Banner"))
        intro.addArranged(Typography.body(
            "EBMediatedBannerAd — 서버가 정한 순서대로 어댑터를 순차 시도하고 " +
            "첫 성공 네트워크의 광고를 노출합니다. autoRefresh는 지원되지 않습니다."
        ))
        stack.addArrangedSubview(intro)

        let status = CardView()
        status.addArranged(SectionLabel.make("Status"))
        status.addArranged(badge)
        status.addArranged(winnerRow)
        stack.addArrangedSubview(status)

        stack.addArrangedSubview(options)

        let actions = CardView()
        actions.addArranged(SectionLabel.make("Action"))
        actions.addArranged(PrimaryButton.filled("Load mediated banner", target: self, action: #selector(handleLoad)))
        stack.addArrangedSubview(actions)

        let hostCard = CardView()
        hostCard.addArranged(SectionLabel.make("Creative"))
        bannerHost.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            bannerHost.heightAnchor.constraint(equalToConstant: 50)
        ])
        hostCard.addArranged(bannerHost)
        stack.addArrangedSubview(hostCard)

        let waterfallCard = CardView()
        waterfallCard.addArranged(SectionLabel.make("Waterfall"))
        waterfallCard.addArranged(waterfallView)
        stack.addArrangedSubview(waterfallCard)

        let logCard = CardView()
        logCard.addArranged(SectionLabel.make("Lifecycle log"))
        logCard.addArranged(logView)
        stack.addArrangedSubview(logCard)

        NSLayoutConstraint.activate([
            scroll.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scroll.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
            scroll.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            scroll.trailingAnchor.constraint(equalTo: view.trailingAnchor),

            stack.topAnchor.constraint(equalTo: scroll.contentLayoutGuide.topAnchor, constant: Spacing.l),
            stack.bottomAnchor.constraint(equalTo: scroll.contentLayoutGuide.bottomAnchor, constant: -Spacing.xxl),
            stack.leadingAnchor.constraint(equalTo: scroll.frameLayoutGuide.leadingAnchor, constant: Spacing.l),
            stack.trailingAnchor.constraint(equalTo: scroll.frameLayoutGuide.trailingAnchor, constant: -Spacing.l)
        ])

        badge.set(.idle, text: "Tap to load")
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        bannerAd?.stop()
    }

    @objc private func handleLoad() {
        let ad = bannerAd ?? makeBannerAd()
        ad.stop()
        options.apply(to: ad.options)
        winnerRow.set("-")
        waterfallView.append("─── load ───")
        set(.loading, "loading…")
        ad.load()
    }

    private func makeBannerAd() -> EBMediatedBannerAd {
        let banner = EBMediatedBannerAd(adUnitId: AdUnitIds.banner,
                                        size: CGSize(width: 320, height: 50))
        banner.perNetworkTimeout = 5.0
        banner.rootViewControllerProvider = { [weak self] in self }
        banner.translatesAutoresizingMaskIntoConstraints = false
        bannerHost.addSubview(banner)
        NSLayoutConstraint.activate([
            banner.centerXAnchor.constraint(equalTo: bannerHost.centerXAnchor),
            banner.centerYAnchor.constraint(equalTo: bannerHost.centerYAnchor),
            banner.widthAnchor.constraint(equalToConstant: 320),
            banner.heightAnchor.constraint(equalToConstant: 50)
        ])

        banner.onWaterfallEvent = { [weak self] event in
            self?.waterfallView.append(WaterfallFormatter.format(event))
        }
        banner.onLoad = { [weak self, weak banner] in
            self?.winnerRow.set(banner?.winningNetwork ?? "-")
            self?.set(.ready, "loaded")
        }
        banner.onFail        = { [weak self] error in self?.set(.failed, "failed: \(error.localizedDescription)") }
        banner.onClick       = { [weak self] in self?.set(.clicked, "clicked") }
        banner.onLeaveApp    = { [weak self] in self?.set(.clicked, "left app") }
        banner.onClickFinish = { [weak self] in self?.set(.ready,   "click finished") }
        bannerAd = banner
        return banner
    }

    private func set(_ state: StatusBadge.State, _ text: String) {
        badge.set(state, text: text)
        logView.append(text)
    }
}
